File: Sheets/src/neetcode/practice/Arrays/_3_twoSum.java
package neetcode.practice.Arrays;

import java.util.HashMap;

class Solution {
  public int[] twoSum(int[] nums, int target) {
    HashMap<Integer, Integer> map = new HashMap<>();
    int n = nums.length;
    for (int i = 0; i < n; i++)
      map.put(nums[i], i);

    for (int i = 0; i < n; i++) {
      int num = nums[i];
      if (map.containsKey(target - num)) {
        int index = map.get(target - num);
        if (index != i)
          return new int[] { Math.min(i, index), Math.max(i, index) };
      }
    }
    return null;
  }
}

File: Sheets/bin/neetcode/practice/Arrays/_3_TwoSum.cc
#include <unordered_map>
#include <vector>

class Solution {
public:
  std::vector<int> twoSum(std::vector<int> &nums, int target) {
    std::unordered_map<int, int> index{};

    for (int i = 0; i < static_cast<int>(nums.size()); i++) {
      int num = nums[i];
      auto it = index.find(target - num);
      if (it != index.end())
        return {it->second, i};
      else
        index.insert({num, i});
    }

    return {};
  }
};
